network/limitlistener/rejectingLimitListener_test.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2013 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package limitlistener_test
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/algorand/go-algorand/network/limitlistener"
+	"github.com/algorand/go-algorand/test/partitiontest"
+)
+
+const defaultMaxOpenFiles = 256
+const timeout = 5 * time.Second
+
+func TestRejectingLimitListener(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	const max = 5
+	attempts := (maxOpenFiles() - max) / 2
+	if attempts > 256 { // maximum length of accept queue is 128 by default
+		attempts = 256
+	}
+
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer l.Close()
+	l = limitlistener.RejectingLimitListener(l, max)
+
+	var open int32
+	go http.Serve(l, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if n := atomic.AddInt32(&open, 1); n > max {
+			t.Errorf("%d open connections, want <= %d", n, max)
+		}
+		defer atomic.AddInt32(&open, -1)
+		time.Sleep(500 * time.Millisecond)
```

Comment:
using these constants guarantee that your test would randomly fail on CI system, where the machine would randomly be super sluggish.
Instead of waiting for a given timeout, use a channel to "lock" this go-routine at this state. That should give you the ability to control the scheduling of each goroutine.

Reply:
I changed the test, but unfortunately, there is no way to completely avoid timing assumptions.